packages/core/src/StateNode.ts

```diff
@@ -1015,17 +1012,20 @@ class StateNode<
     prevState?: State<TContext, any, any, any, any>,
     predictableExec?: PredictableActionArgumentsExec
   ): Array<Array<ActionObject<TContext, TEvent>>> {
-    const prevConfig = getConfiguration(
-      [],
-      prevState ? this.getStateNodes(prevState.value) : [this]
-    );
+    const prevConfig = prevState
+      ? getConfiguration([], this.getStateNodes(prevState.value))
+      : [];
+    const entrySet = new Set<StateNode<any, any, any, any, any, any>>();
 
-    for (const sn of resolvedConfig) {
+    for (const sn of Array.from(resolvedConfig).sort(
+      (a, b) => a.order - b.order
+    )) {
       if (
         !has(prevConfig, sn) ||
-        (has(transition.entrySet, sn.parent) && !has(transition.entrySet, sn))
+        has(transition.exitSet, sn) ||
+        (sn.parent && entrySet.has(sn.parent))
       ) {
-        transition.entrySet.push(sn);
+        entrySet.add(sn);
       }
     }
     for (const sn of prevConfig) {
```

Comment:
I think that the logic in this for-of loop is fragile but I couldn't make it break right now. Perhaps it's because nodes in the `prevConfig` are already sorted in a way that allows this for-of loop to do its job.